README.md
# flutter_profile_card

This plugin allows you to display a user profile with its followers and followings.

Also allows to have a button to follow the user profile.

![](https://media.giphy.com/media/SbxjWP7WmC4EmAPcfD/giphy.gif)

## Getting Started

```
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_profile_card/flutter_profile_card.dart';
import 'package:flutter_profile_card/models/Profile.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool _isFollow = false;
  ProfileModel _profileModel;

  @override
  void initState() {
    // Define the model
    _profileModel = ProfileModel(
      uid: '1234567890',
      username: 'John Doe',
      status: 'John doe is certainly the best development but very discreet...',
      avatarURL: 'https://picsum.photos/200',
      followers: 1,
      followings: 1234,
    );

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        accentColor: Colors.green,
        textTheme: TextTheme(
          bodyText2: TextStyle(
            fontSize: 16.0,
          ),
        ),
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: ProfileCard(
          // Add the model in the widget parameter
          profile: _profileModel,
          // Capture the follow event
          onFollowed: (ProfileModel profile) {
            print(profile.toJson());
            setState(() {
              // Database code...
              _isFollow = !_isFollow;

              _profileModel.isFollowed = _isFollow;
              _profileModel.updateFollowers = _isFollow ? 2 : 1;
            });
          },
        ),
      ),
    );
  }
}

```


CHANGELOG.md
## 0.0.3+1

- Optimized follow button on small screens

## 0.0.3

- Added loading state when the user click on the button follow

## 0.0.2+3

- Updated the version of Flutter Models

## 0.0.2+2

- Added Hero class on Avatar Profile

## 0.0.2

- Included the package flutter models [flutter_models](https://pub.dev/packages/flutter_models)
- Updated documentation

## 0.0.2

- Included the package profile avatar [flutter_profile_avatar](https://pub.dev/packages/flutter_profile_avatar)

## 0.0.1

- Initial version


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_profile_card_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/FlutterModelCard_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_profile_card/flutter_profile_card.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:image_test_utils/image_test_utils.dart';
import 'package:flutter_models/models/UserModel.dart';

main() {
  testWidgets('Should see basics componants', (WidgetTester tester) async {
    // ARRANGE
    UserModel profileModel = UserModel(
      uid: '123456789',
      username: 'John Doe',
      status: 'status',
      avatarURL: 'status',
      followers: 2,
      followings: 1,
    );

    await provideMockedNetworkImages(() async {
      await tester.pumpWidget(
        MaterialApp(
          title: 'Flutter Demo',
          home: ProfileCard(
            profile: profileModel,
          ),
        ),
      );
    });

    // ACT
    final Finder usernameFinder =
        find.text(profileModel.username.toUpperCase());
    final Finder followersNumberFinder =
        find.text(profileModel.followers.toString());
    final Finder followingNumberFinder =
        find.text(profileModel.followings.toString());
    final Finder followButton = find.text('follow'.toUpperCase());

    // ASSERT
    expect(usernameFinder, findsOneWidget);
    expect(followersNumberFinder, findsOneWidget);
    expect(followingNumberFinder, findsOneWidget);
    expect(followButton, findsNothing);
  });

  testWidgets('Should see follow button', (WidgetTester tester) async {
    // ARRANGE
    UserModel profileModel = UserModel(
      uid: '123456789',
      username: 'John Doe',
      status: 'status',
      avatarURL: 'status',
      followers: 2,
      followings: 1,
    );

    await provideMockedNetworkImages(() async {
      await tester.pumpWidget(
        MaterialApp(
          title: 'Flutter Demo',
          home: ProfileCard(
            profile: profileModel,
            onFollowed: (_) => null,
          ),
        ),
      );
    });

    // ACT
    final Finder followButton = find.text('follow'.toUpperCase());

    // ASSERT
    expect(followButton, findsOneWidget);
  });

  testWidgets('Should see unfollow button', (WidgetTester tester) async {
    // ARRANGE
    UserModel profileModel = UserModel(
      uid: '123456789',
      username: 'John Doe',
      status: 'status',
      avatarURL: 'status',
      followers: 2,
      followings: 1,
      isFollow: true,
    );

    await provideMockedNetworkImages(() async {
      await tester.pumpWidget(
        MaterialApp(
          title: 'Flutter Demo',
          home: ProfileCard(
            profile: profileModel,
            onFollowed: (_) => null,
          ),
        ),
      );
    });

    // ACT
    final Finder followButton = find.text('unfollow'.toUpperCase());

    // ASSERT
    expect(followButton, findsOneWidget);
  });

  testWidgets('Should capture follow button event',
      (WidgetTester tester) async {
    // ARRANGE
    bool event = false;
    UserModel profileModel = UserModel(
      uid: '123456789',
      username: 'John Doe',
      status: '',
      avatarURL: 'status',
      followers: 2,
      followings: 1,
    );

    await provideMockedNetworkImages(() async {
      await tester.pumpWidget(
        MaterialApp(
          title: 'Flutter Demo',
          home: ProfileCard(
            profile: profileModel,
            onFollowed: (_) => event = true,
          ),
        ),
      );
    });

    // ACT
    final Finder followButton = find.text('follow'.toUpperCase());
    await tester.tap(followButton);

    // ASSERT
    expect(event, true);
  });

  testWidgets('Should not see about section', (WidgetTester tester) async {
    // ARRANGE
    UserModel profileModel = UserModel(
      uid: '123456789',
      username: 'John Doe',
      status: '',
      avatarURL: 'status',
      followers: 2,
      followings: 1,
      isFollow: true,
    );

    await provideMockedNetworkImages(() async {
      await tester.pumpWidget(
        MaterialApp(
          title: 'Flutter Demo',
          home: ProfileCard(
            profile: profileModel,
            onFollowed: (_) => null,
          ),
        ),
      );
    });

    // ACT
    final Finder aboutSection = find.text('about'.toUpperCase());

    // ASSERT
    expect(aboutSection, findsNothing);
  });
}


lib/widgets/Stat.dart
import 'package:flutter/material.dart';

class StatWidget extends StatelessWidget {
  final String number;
  final String label;

  const StatWidget({
    Key key,
    @required this.number,
    @required this.label,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Text(
          number,
          style: TextStyle(
            fontSize: Theme.of(context).textTheme.headline4.fontSize,
            fontWeight: Theme.of(context).textTheme.headline4.fontWeight,
            color: Colors.white,
          ),
        ),
        Text(
          label.toUpperCase(),
          style: TextStyle(
            fontSize: Theme.of(context).textTheme.headline6.fontSize,
            fontWeight: Theme.of(context).textTheme.headline4.fontWeight,
            color: Colors.white,
          ),
        ),
      ],
    );
  }
}


lib/flutter_profile_card.dart
import 'package:flutter/material.dart';
import 'package:flutter_profile_avatar/flutter_profile_avatar.dart';
import 'package:flutter_profile_card/widgets/Stat.dart';
import 'package:flutter_models/models/UserModel.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';

class ProfileCard extends StatefulWidget {
  final UserModel profile;
  final Function(UserModel) onFollowed;

  final String aboutLabel;
  final String followLabel;
  final String unfollowLabel;
  final String followersLabel;
  final String followingsLabel;

  const ProfileCard({
    Key key,
    @required this.profile,
    this.onFollowed,
    this.aboutLabel = 'About',
    this.followLabel = 'Follow',
    this.unfollowLabel = 'Unfollow',
    this.followersLabel = 'Followers',
    this.followingsLabel = 'Followings',
  }) : super(key: key);

  @override
  _ProfileCardState createState() => _ProfileCardState();
}

class _ProfileCardState extends State<ProfileCard> {
  bool _loading = false;

  @override
  Widget build(BuildContext context) {
    return ListView(
      children: [
        Container(
          padding: const EdgeInsets.symmetric(
            vertical: 20.0,
          ),
          color: Theme.of(context).primaryColor,
          child: Column(
            children: [
              Padding(
                padding: const EdgeInsets.only(
                  bottom: 20.0,
                ),
                child: Container(
                  decoration: BoxDecoration(
                    shape: BoxShape.circle,
                    boxShadow: [
                      BoxShadow(
                        blurRadius: 3,
                        color: Colors.grey[700],
                        offset: Offset(0, 2),
                      )
                    ],
                  ),
                  child: Hero(
                    tag: widget.profile.uid,
                    child: ProfileAvatar(
                      size: 70,
                      username: widget.profile.username,
                      avatarURL: widget.profile.avatarURL,
                    ),
                  ),
                ),
              ),
              Text(
                widget.profile.username.toUpperCase(),
                style: Theme.of(context).textTheme.headline4.copyWith(
                      color: Colors.white,
                    ),
              ),
              Padding(
                padding: const EdgeInsets.only(
                  top: 15.0,
                ),
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  children: [
                    StatWidget(
                      number: widget.profile.followers.toString(),
                      label: widget.followersLabel,
                    ),
                    StatWidget(
                      number: widget.profile.followings.toString(),
                      label: widget.followingsLabel,
                    )
                  ],
                ),
              )
            ],
          ),
        ),
        Visibility(
          visible: widget.profile.status.isNotEmpty,
          child: Padding(
            padding: const EdgeInsets.only(
              top: 10.0,
              left: 20.0,
              right: 20.0,
            ),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Padding(
                  padding: const EdgeInsets.symmetric(
                    vertical: 10.0,
                  ),
                  child: Text(
                    widget.aboutLabel.toUpperCase(),
                    style: TextStyle(
                      fontWeight: FontWeight.bold,
                      fontSize: Theme.of(context).textTheme.headline6.fontSize,
                      letterSpacing: 1.2,
                    ),
                  ),
                ),
                Text(
                  widget.profile.status,
                  style: TextStyle(
                      fontSize: Theme.of(context).textTheme.bodyText2.fontSize),
                ),
              ],
            ),
          ),
        ),
        Visibility(
          visible: widget.onFollowed != null,
          child: Container(
            padding: const EdgeInsets.symmetric(
              vertical: 30.0,
            ),
            child: Center(
              child: Container(
                width: MediaQuery.of(context).size.width - 60.0,
                child: RaisedButton(
                  padding: const EdgeInsets.symmetric(
                    vertical: 15.0,
                  ),
                  color: !widget.profile.isFollow
                      ? Theme.of(context).primaryColor
                      : Colors.grey,
                  onPressed: () async {
                    setState(() => _loading = !_loading);

                    await widget.onFollowed(widget.profile);

                    setState(() => _loading = !_loading);
                  },
                  child: !_loading
                      ? Text(
                          !widget.profile.isFollow
                              ? widget.followLabel.toUpperCase()
                              : widget.unfollowLabel.toUpperCase(),
                          style: Theme.of(context).textTheme.headline5.copyWith(
                                color: Colors.white,
                              )

                          // TextStyle(
                          //   fontSize:
                          //       Theme.of(context).textTheme.headline5.fontSize,
                          //   color: Colors.white,
                          // ),
                          )
                      : SizedBox(
                          height:
                              Theme.of(context).textTheme.headline5.fontSize +
                                  4,
                          child: SpinKitThreeBounce(
                            color: Colors.white,
                            size: 20.0,
                          ),
                        ),
                ),
              ),
            ),
          ),
        ),
      ],
    );
  }
}


android/src/main/java/com/nicodevelop/flutter_profile_card/FlutterProfileCardPlugin.java
package com.nicodevelop.flutter_profile_card;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterProfileCardPlugin */
public class FlutterProfileCardPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_profile_card");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:flutter_profile_card/flutter_profile_card.dart';
import 'package:flutter_mobile_camera/CameraBuilder.dart';
import 'package:flutter_models/models/UserModel.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  bool _isFollow = false;
  UserModel _profileModel;

  @override
  void initState() {
    // Define the model
    _profileModel = UserModel(
      uid: '1234567890',
      username: 'John Doe',
      status: 'John doe is certainly the best development but very discreet...',
      avatarURL: 'https://picsum.photos/200',
      followers: 1,
      followings: 1234,
    );

    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return CameraBuilder(
      child: MaterialApp(
        theme: ThemeData(
          accentColor: Colors.green,
          textTheme: TextTheme(
            bodyText2: TextStyle(
              fontSize: 16.0,
            ),
          ),
        ),
        home: Scaffold(
          appBar: AppBar(
            title: const Text('Plugin example app'),
          ),
          body: ProfileCard(
            // Add the model in the widget parameter
            profile: _profileModel,
            // Capture the follow event
            onFollowed: (UserModel profile) async {
              print(profile.toJson());

              await Future.delayed(Duration(seconds: 3));

              setState(() {
                // Database code...
                _isFollow = !_isFollow;

                _profileModel.isFollowed = _isFollow;
                _profileModel.updateFollowers = _isFollow ? 2 : 1;
              });
            },
          ),
        ),
      ),
    );
  }
}
